keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1086 @@
+- `null` (allowed during the v1.7 release)
+- `true` (equal in privilege to a v1.7 pod that didn't set the field)
+- `false` (strictly less privileged than other allowed values)
+
+<<[UNRESOLVED]>>
+
+_Blocking for Beta._
+
+How long will old profiles be kept for? What is the removal policy?
+
+<<[/UNRESOLVED]>>
+
+### PodTemplate Resources
+
+Audit and Warn modes are also checked on resource types that embed a PodTemplate (enumerated below),
+but enforce mode only applies to actual pod resources.
+
+Since users do not create pods directly in the typical deployment model, the warning mechanism is
+only effective if it can also warn on templated pod resources. Similarly, for audit it is useful to
+tie the audited violation back to the requesting user, so audit will also apply to templated pod
+resources. In the interest of supporting mutating admission controllers, policies will only
+be enforced on actual pods.
+
+Templated pod resources include:
+
+- v1 ReplicationController
+- v1 PodTemplate
+- apps/v1 ReplicaSet
+- apps/v1 Deployment
+- apps/v1 StatefulSet
+- apps/v1 DaemonSet
+- batch/v1 CronJob
+- batch/v1 Job
+
+PodTemplate warnings & audit will only be applied to built-in types. CRDs that wish to take
+advantage of this functionality can use an object reference to a v1/PodTemplate resource rather than
+inlining a PodTemplate. We will publish a guide (documentation and/or examples) that demonstrate
+this pattern. Alternatively, the functionality can be implemented in a 3rd party admission plugin
+leveraging the library implementation.
+
+### Namespace policy update warnings
+
+When an `enforce` policy (or version) label is added or changed, the admission plugin will test each pod
+in the namespace against the new policy. Violations are returned to the user as warnings. These
+checks have a timeout of XX seconds and a limit of YY pods, and will return a warning in the event
+that not every pod was checked. User exemptions are ignored by these checks, but runtime class
+exemptions still apply. Namespace exemptions are also ignored, but an additional warning will be
+returned when updating the policy on an exempt namespace. These checks only consider actual Pod
+resources, not [templated pods].
+
+These checks are also performed when making a dry-run request, which can be an effective way of
+checking for breakages before updating a policy, for example:
+
+```
+kubectl label --dry-run=server --overwrite ns --all pod-security.kubernetes.io/enforce=baseline
+```
+
+<<[UNRESOLVED]>>
+
+_Non-blocking: can be decided on the implementing PR_
+
+- What should the timeout be for pod update warnings?
+  - Total is a parameter on the context (query parameter for webhooks). Cap should be
+    `min(timeout_param, hard_cap)`, where the `hard_cap` is a small number of seconds.
+  - Expect evaluation to be fast, so even 3k pods should come in well under the timeout.
+- What should the pod limit be set to?
+  - 3,000 is the
```

Comment:
For namespaces with more than 3k pods, is the set we check for violations deterministic? Or will we get random subsets on consecutive dry runs?

Reply:
I'm not sure... the answer might be different for the built-in and webhook implementations. If we have the full list of pods, we could do a sort before dropping pods to keep it deterministic. We can also reevaluate this number once we have benchmarks.